## Business Problem:

The company  is expanding in to new industries to diversify its portfolio. Specifically, they are interested in purchasing and operating airplanes for commercial and private enterprises, but do not know anything about the potential risks of aircraft. You are charged with determining which aircraft are the lowest risk for the company to start this new business endeavor. You must then translate your findings into actionable insights that the head of the new aviation division can use to help decide which aircraft to purchase.

## Aim and Goals:

This project analyzes aviation accident data from 1962 to 2023 to identify the safest aircraft models. By understanding accident trends, risk factors, and aircraft performance, we aim to provide actionable insights that guide the company in selecting low-risk aircraft for its new aviation division.
Project Goals:
-  Identify aircraft models with the lowest accident rates.
-  Highlight key risk factors contributing to aviation accidents.
-  Provide data-backed recommendations for safe aircraft investme


## 1. Understanding the data

Understanding the data helps to clearly grasp what the dataset entails, including what each column represents, the types of values contained, and how complete or consistent it is. This clarity allows you to identify relevant patterns, spot missing or incorrect entries, to ultimately ensures that you conduct accurate and meaningful analyses.

In [160]:
# loading the data
import pandas as pd 
aviationdata = pd.read_csv("Aviation_Data /AviationData.csv", encoding='cp1252')

/var/folders/3s/d5pt38l9793djb6wv9n_4d6w0000gn/T/ipykernel_76904/3486649267.py:3: DtypeWarning: Columns (6,7,28) have mixed types. Specify dtype option on import or set low_memory=False.
  aviationdata = pd.read_csv("Aviation_Data /AviationData.csv", encoding='cp1252')


The warning means that pandas found columns containing mixed data types (strings and numbers).
It's a common warning with large or unstructured datasets which now brings us to the next step of cleaning the data

In [161]:
# viewing the first five rows
aviationdata.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude  Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN        NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN        NaN          NaN   
2    Saltville, VA  United States  36.922223 -81.878056          NaN   
3       EUREKA, CA  United States        NaN        NaN          NaN   
4       Canton, OH  United States        NaN        NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

In [162]:
# viewing the general info of the data 
aviationdata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50132 non-null  object 
 9   Airport.Name            52704 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87507 non-null  object 
 14  Make                    88826 non-null

The above general infomation helps to quickly understand the data I am working on in my case:
- The dataset contains 88,889 entries and 31 columns.
- Reveals the type of data if its text ,numbers or floats in each column.
- Identifies columns with missing data.
- Reveals the current data type of data each column.

# 2. Data Cleaning 

Data cleaning is crucial  as it ensures our data is clean, structured, and reliable to perform accurate analysis and extract meaningful insights.
This process will include:
- Handling missing values
- Dealing with duplicates
- Correcting data types
- standardizing the date formarts

In [257]:
''' Creating a copy of my data so as to work on the copy instead of of original data'''
avidatacopy = aviationdata.copy()

In [258]:
avidatacopy.head(n=3)

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   

          Location        Country   Latitude  Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN        NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN        NaN          NaN   
2    Saltville, VA  United States  36.922223 -81.878056          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  

[3 rows x 31 columns]

## i) Handling missing values 

In [259]:
# Checking for all the columns with missing values and the count of values missing
avidatacopy.isnull().sum()[avidatacopy.isnull().sum()>0]

Location                     52
Country                     226
Latitude                  54507
Longitude                 54516
Airport.Code              38757
Airport.Name              36185
Injury.Severity            1000
Aircraft.damage            3194
Aircraft.Category         56602
Registration.Number        1382
Make                         63
Model                        92
Amateur.Built               102
Number.of.Engines          6084
Engine.Type                7096
FAR.Description           56866
Schedule                  76307
Purpose.of.flight          6192
Air.carrier               72241
Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
Weather.Condition          4492
Broad.phase.of.flight     27165
Report.Status              6384
Publication.Date          13771
dtype: int64

- Replacing the Location missing values with "Unknown" since we do not know the location names

In [260]:
# Handling missing values in the Location column 
avidatacopy["Location"]= avidatacopy["Location"].fillna("Unknown")

In [261]:
avidatacopy["Location"].isnull().sum()

0

- Filling the missing country names with the most frequent country occuring 

In [262]:
#The most frequent country is United States
avidatacopy["Country"].value_counts()

Country
United States                       82248
Brazil                                374
Canada                                359
Mexico                                358
United Kingdom                        344
                                    ...  
Seychelles                              1
Palau                                   1
Libya                                   1
Saint Vincent and the Grenadines        1
Turks and Caicos Islands                1
Name: count, Length: 219, dtype: int64

In [263]:
#filling the missing values with the mode()
frequent_country = avidatacopy["Country"].mode().iloc[0]
frequent_country

'United States'

In [264]:
# replacing the missing values withe mode()
avidatacopy["Country"] = avidatacopy["Country"].fillna(frequent_country)

In [265]:
avidatacopy["Country"].isnull().sum()

0

- Dropping both the Latitude and longitude columns since they have alot of missing values 
- In my analysis I will use the Locations and country instead of the latitude and longitude

In [266]:
#Dropping both the latitude and longitude columns
avidatacopy.drop(columns=['Latitude', 'Longitude'], inplace=True)

- Dropping the Airport.Code column for it has alot of missing values

In [267]:
avidatacopy["Airport.Code"].value_counts()

Airport.Code
NONE    1488
PVT      485
APA      160
ORD      149
MRI      137
        ... 
7NJ9       1
CWV        1
5QA        1
M55        1
EIKH       1
Name: count, Length: 10374, dtype: int64

In [268]:
avidatacopy = avidatacopy.drop(columns = ["Airport.Code"])

In [269]:
# Checking the remaining columns after dropping some of them
avidatacopy.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Airport.Name', 'Injury.Severity',
       'Aircraft.damage', 'Aircraft.Category', 'Registration.Number', 'Make',
       'Model', 'Amateur.Built', 'Number.of.Engines', 'Engine.Type',
       'FAR.Description', 'Schedule', 'Purpose.of.flight', 'Air.carrier',
       'Total.Fatal.Injuries', 'Total.Serious.Injuries',
       'Total.Minor.Injuries', 'Total.Uninjured', 'Weather.Condition',
       'Broad.phase.of.flight', 'Report.Status', 'Publication.Date'],
      dtype='object')

- filling the Airport.Name missing values with "Unknown" since we were not given the Airport.Names

In [270]:
avidatacopy["Airport.Name"] = avidatacopy["Airport.Name"].fillna("Unknown")

In [271]:
# Confirming the data information so far after the changes.
avidatacopy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 28 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88889 non-null  object 
 5   Country                 88889 non-null  object 
 6   Airport.Name            88889 non-null  object 
 7   Injury.Severity         87889 non-null  object 
 8   Aircraft.damage         85695 non-null  object 
 9   Aircraft.Category       32287 non-null  object 
 10  Registration.Number     87507 non-null  object 
 11  Make                    88826 non-null  object 
 12  Model                   88797 non-null  object 
 13  Amateur.Built           88787 non-null  object 
 14  Number.of.Engines       82805 non-null

Dropping the numbers in brackets after Fatal to categorize all the fatal as "Fatal" and filling the missing values with the mode().

In [272]:
#viewing the column to understand how it looks like
avidatacopy["Injury.Severity"].value_counts()

Injury.Severity
Non-Fatal     67357
Fatal(1)       6167
Fatal          5262
Fatal(2)       3711
Incident       2219
              ...  
Fatal(270)        1
Fatal(60)         1
Fatal(43)         1
Fatal(143)        1
Fatal(230)        1
Name: count, Length: 109, dtype: int64

In [275]:
# Dropping the numbers in brackets after the Fatal and accounting for different values in the column including NaN
def cleaned_injury_severity(data):
    cleaned_data = []
    for word in data:
        if isinstance(word, float):
            cleaned_data.append(None)
        elif word == "Fatal":
            cleaned_data.append(word)
        elif "Fatal" in word:
            cleaned_data.append(word.split("(")[0])
        else:
            cleaned_data.append(word)
    return cleaned_data

avidatacopy["Injury.Severity"] = cleaned_injury_severity(avidatacopy["Injury.Severity"])

In [276]:
#Confirming the code has changed and worked
avidatacopy["Injury.Severity"].value_counts()

Injury.Severity
Non-Fatal      67357
Fatal          17826
Incident        2219
Minor            218
Serious          173
Unavailable       96
Name: count, dtype: int64

In [278]:
# Checking the missing values 
avidatacopy["Injury.Severity"].isnull().sum()

1000

- We have 1000 missing values from the avidatacopy["Injury.Severity"]. 
- To make sure the values are uniform I will fill the missing values with "Unavailable" since we were not given the severity of the accident.
- Incident is not clear whether injury occured and was not recorded or there was no injury *It required further investigation*
- "Incidents" in the column is meant to represent injury severity, but it suggests that either the data entry is inconsistent or that the information is being misclassified 
- Non-Fatal does not indicate whether the injuries were serious or minor ,it only conculdes that the accident did not lead to death.
- To investigate this column I will use the Total.Fatal.Injuries ,Total.Serious.Injuries ,Total.Minor.Injuries ,Total.Uninjured  columns to determine the Injury .Severity.
- I will create another column to clasiffy the Severity as [Fatal,Serious injuries,Minor injuries,Uninjured,Unknown] This will give a clear view of the severity while clasiffying even the Unknown ones.

To achieve the above I first have to ensure the columns ["Total.Fatal.Injuries","Total.Serious.Injuries","Total.Minor.Injuries","Total.Uninjured"]
Do not have missing values and if they have I will use the placeholder of "Unknow" to avoiding assuming the counts 

In [285]:
# Checking for missing values in the column Total.Fatal.Injuries
injury_cols = ['Total.Fatal.Injuries', 'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured']
missing_value_sum = avidatacopy[injury_cols].isnull().sum()
missing_value_sum

Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
dtype: int64

The above code prints the sum of all the missing values in the columns I am working with.To ensure I get the expected results the missing values will be filled with "Unknown"